gift-certificate-data/src/main/java/com/epam/esm/giftcertificatemodule3/dao/impl/GiftCertificateDAOHibernate.java

```diff
@@ -2,54 +2,128 @@
 
 import com.epam.esm.giftcertificatemodule3.dao.GiftCertificateDAO;
 import com.epam.esm.giftcertificatemodule3.entity.GiftCertificate;
-import org.hibernate.Session;
-import org.hibernate.query.Query;
-import org.springframework.beans.factory.annotation.Autowired;
+import com.epam.esm.giftcertificatemodule3.entity.Tag;
+import com.epam.esm.giftcertificatemodule3.model.SearchParametersHolder;
 import org.springframework.stereotype.Repository;
 
 import javax.persistence.EntityManager;
+import javax.persistence.PersistenceContext;
+import javax.persistence.Query;
+import javax.persistence.TypedQuery;
+import javax.persistence.criteria.*;
+import java.time.ZonedDateTime;
+import java.util.ArrayList;
 import java.util.List;
 
 @Repository
 public class GiftCertificateDAOHibernate implements GiftCertificateDAO {
 
+    @PersistenceContext
     private EntityManager entityManager;
 
-    @Autowired
-    public GiftCertificateDAOHibernate(EntityManager entityManager) {
-        this.entityManager = entityManager;
+    @Override
+    public void save(GiftCertificate giftCertificate) {
+        giftCertificate.setCreateDate(ZonedDateTime.now().toOffsetDateTime());
+        giftCertificate.setLastUpdateDate(ZonedDateTime.now().toOffsetDateTime());
+        GiftCertificate merged = entityManager.merge(giftCertificate);
+        giftCertificate.setId(merged.getId());
+        giftCertificate.setTags(merged.getTags());
     }
 
     @Override
-    public List<GiftCertificate> findAll() {
-        Session session = entityManager.unwrap(Session.class);
```

Comment:
Hibernate should be used as a JPA implementation for data access. Why do you use Session instead of EntityManager?